storage/watchable_store.go

```diff
@@ -241,57 +243,133 @@ func (s *watchableStore) syncWatchingsLoop() {
 	}
 }
 
-// syncWatchings syncs the watchings in the unsyncd map.
-func (s *watchableStore) syncWatchings() {
+// rangeUnsynced ranges on all unsynced watchings and returns all
+// revision bytes, key-value pairs and next revision.
+func (s *watchableStore) rangeUnsynced() ([][]byte, []storagepb.KeyValue, int64, map[string]map[*watching]struct{}, error) {
+	// get minimum and maximum revision to get all key-value pairs in unsynced
+	minRev, maxRev := int64(math.MaxInt64), int64(math.MinInt64)
+	keyToWatchings := make(map[string]map[*watching]struct{})
 	_, curRev, _ := s.store.Range(nil, nil, 0, 0)
 	for w := range s.unsynced {
-		var end []byte
-		if w.prefix {
-			end = make([]byte, len(w.key))
-			copy(end, w.key)
-			end[len(w.key)-1]++
-		}
-		limit := cap(w.ch) - len(w.ch)
-		// the channel is full, try it in the next round
-		if limit == 0 {
+		if w.cur > 0 && w.cur <= s.store.compactMainRev {
+			// TODO: handle this ErrCompacted error
+			log.Printf("storage: %v", ErrCompacted)
+			delete(s.unsynced, w)
 			continue
 		}
-		revbs, kvs, nextRev, err := s.store.RangeHistory(w.key, end, int64(limit), w.cur)
-		if err != nil {
-			// TODO: send error event to watching
+		if w.cur > s.store.currentRev.main {
+			// TODO: handle this ErrFutureRev error
```

Comment:
A watching becomes unsynced when its chan buffer is full and it is behind current rev. If a watching's current is lager than current rev, how does it become unsynced at the first place? We do not need to handle the error. We should simply panic here.